ecommerce/api.py

```diff
@@ -0,0 +1,164 @@
+"""
+Functions for ecommerce
+"""
+from base64 import b64encode
+from datetime import datetime
+import hashlib
+import hmac
+import logging
+import uuid
+
+from django.conf import settings
+from django.db import transaction
+from django.http.response import Http404
+from django.shortcuts import get_object_or_404
+from rest_framework.exceptions import ValidationError
+
+from courses.models import CourseRun
+from ecommerce.models import (
+    Line,
+    Order,
+)
+from profiles.api import get_social_username
+
+
+ISO_8601_FORMAT = '%Y-%m-%dT%H:%M:%SZ'
+log = logging.getLogger(__name__)
+
+
+def get_purchasable_course_run(course_key, user):
+    """
+    Gets a course run, or raises Http404 if not purchasable
+
+    Args:
+        course_key (str):
+            An edX course key
+        user (User):
+            The purchaser of the course run
+    Returns:
+        CourseRun: A course run
+    """
+    # Make sure it's connected to a live program, it has a valid price, and the user is enrolled in the program already
+    try:
+        course_run = get_object_or_404(
+            CourseRun,
+            edx_course_key=course_key,
+            course__program__live=True,
+            courseprice__is_valid=True,
+        )
+    except Http404:
+        log.warning("Course run %s is not purchasable", course_key)
+        raise
+
+    # Make sure it's not already purchased
+    if Line.objects.filter(
+            order__status=Order.FULFILLED,
+            order__user=user,
+            course_key=course_run.edx_course_key,
+    ).exists():
+        log.warning("Course run %s is already purchased by user %s", course_key, user)
+        raise ValidationError("Course run {} is already purchased".format(course_key))
+
+    return course_run
+
+
+@transaction.atomic
+def create_unfulfilled_order(course_id, user):
+    """
+    Create a new Order which is not fulfilled for a purchasable course run. If course run is not purchasable,
+    it raises an Http404
+
+    Args:
+        course_id (str):
+            A course key
+        user (User):
+            The purchaser of the course run
+    Returns:
+        Order: A newly created Order for the CourseRun with the given course_id
+    """
+    course_run = get_purchasable_course_run(course_id, user)
+    price = course_run.courseprice_set.get(is_valid=True).price
+    order = Order.objects.create(
+        status=Order.CREATED,
+        total_price_paid=price,
+        user=user,
+    )
+    Line.objects.create(
+        order=order,
+        course_key=course_run.edx_course_key,
+        description='Seat for {}'.format(course_run.title),
+        price=price,
+    )
```

Comment:
More a general comment than a request for change: I noticed that edX makes extensive use of factories even in the normal workflow and not only in the tests. This would be the perfect usage: you will call the factory for `Line` and then you override the `.create` method to take care of creating the order first.
Again, only food for thoughts.